src/shadow-4.1.4/man/login.1
'\" t
.\"     Title: login
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.74.3 <http://docbook.sf.net/>
.\"      Date: 05/10/2009
.\"    Manual: User Commands
.\"    Source: User Commands
.\"  Language: English
.\"
.TH "LOGIN" "1" "05/10/2009" "User Commands" "User Commands"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
login \- begin session on the system
.SH "SYNOPSIS"
.HP \w'\fBlogin\fR\ 'u
\fBlogin\fR [\-p] [\-h\ \fIhost\fR] [\fIusername\fR] [\fIENV=VAR\fR...]
.HP \w'\fBlogin\fR\ 'u
\fBlogin\fR [\-p] [\-h\ \fIhost\fR] \-f \fIusername\fR
.HP \w'\fBlogin\fR\ 'u
\fBlogin\fR [\-p] \-r\ \fIhost\fR
.SH "DESCRIPTION"
.PP
The
\fBlogin\fR
program is used to establish a new session with the system\&. It is normally invoked automatically by responding to the
\fIlogin:\fR
prompt on the user\'s terminal\&.
\fBlogin\fR
may be special to the shell and may not be invoked as a sub\-process\&. When called from a shell,
\fBlogin\fR
should be executed as
\fBexec login\fR
which will cause the user to exit from the current shell (and thus will prevent the new logged in user to return to the session of the caller)\&. Attempting to execute
\fBlogin\fR
from any shell but the login shell will produce an error message\&.
.PP
The user is then prompted for a password, where appropriate\&. Echoing is disabled to prevent revealing the password\&. Only a small number of password failures are permitted before
\fBlogin\fR
exits and the communications link is severed\&.
.PP
If password aging has been enabled for your account, you may be prompted for a new password before proceeding\&. You will be forced to provide your old password and the new password before continuing\&. Please refer to
\fBpasswd\fR(1)
for more information\&.
.PP
After a successful login, you will be informed of any system messages and the presence of mail\&. You may turn off the printing of the system message file,
/etc/motd, by creating a zero\-length file
\&.hushlogin
in your login directory\&. The mail message will be one of "\fIYou have new mail\&.\fR", "\fIYou have mail\&.\fR", or "\fINo Mail\&.\fR" according to the condition of your mailbox\&.
.PP
Your user and group ID will be set according to their values in the
/etc/passwd
file\&. The value for
\fB$HOME\fR,
\fB$SHELL\fR,
\fB$PATH\fR,
\fB$LOGNAME\fR, and
\fB$MAIL\fR
are set according to the appropriate fields in the password entry\&. Ulimit, umask and nice values may also be set according to entries in the GECOS field\&.
.PP
On some installations, the environmental variable
\fB$TERM\fR
will be initialized to the terminal type on your tty line, as specified in
/etc/ttytype\&.
.PP
An initialization script for your command interpreter may also be executed\&. Please see the appropriate manual section for more information on this function\&.
.PP
A subsystem login is indicated by the presence of a "*" as the first character of the login shell\&. The given home directory will be used as the root of a new file system which the user is actually logged into\&.
.PP
The
\fBlogin\fR
program is NOT responsible for removing users from the utmp file\&. It is the responsibility of
\fBgetty\fR(8)
and
\fBinit\fR(8)
to clean up apparent ownership of a terminal session\&. If you use
\fBlogin\fR
from the shell prompt without
\fBexec\fR, the user you use will continue to appear to be logged in even after you log out of the "subsession"\&.
.SH "OPTIONS"
.PP
\fB\-f\fR
.RS 4
Do not perform authentication, user is preauthenticated\&.
.sp
Note: In that case,
\fIusername\fR
is mandatory\&.
.RE
.PP
\fB\-h\fR
.RS 4
Name of the remote host for this login\&.
.RE
.PP
\fB\-p\fR
.RS 4
Preserve environment\&.
.RE
.PP
\fB\-r\fR
.RS 4
Perform autologin protocol for rlogin\&.
.RE
.PP
The
\fB\-r\fR,
\fB\-h\fR
and
\fB\-f\fR
options are only used when
\fBlogin\fR
is invoked by root\&.
.SH "CAVEATS"
.PP
This version of
\fBlogin\fR
has many compilation options, only some of which may be in use at any particular site\&.
.PP
The location of files is subject to differences in system configuration\&.
.PP
The
\fBlogin\fR
program is NOT responsible for removing users from the utmp file\&. It is the responsibility of
\fBgetty\fR(8)
and
\fBinit\fR(8)
to clean up apparent ownership of a terminal session\&. If you use
\fBlogin\fR
from the shell prompt without
\fBexec\fR, the user you use will continue to appear to be logged in even after you log out of the "subsession"\&.
.PP
As with any program,
\fBlogin\fR\'s appearance can be faked\&. If non\-trusted users have physical access to a machine, an attacker could use this to obtain the password of the next person coming to sit in front of the machine\&. Under Linux, the SAK mechanism can be used by users to initiate a trusted path and prevent this kind of attack\&.
.SH "CONFIGURATION"
.PP
The following configuration variables in
/etc/login\&.defs
change the behavior of this tool:
.PP
\fBCONSOLE\fR (string)
.RS 4
If defined, either full pathname of a file containing device names or a ":" delimited list of device names\&. Root logins will be allowed only upon these devices\&.
.sp
If not defined, root will be allowed on any device\&.
.RE
.PP
\fBCONSOLE_GROUPS\fR (string)
.RS 4
List of groups to add to the user\'s supplementary groups set when logging in on the console (as determined by the CONSOLE setting)\&. Default is none\&.

Use with caution \- it is possible for users to gain permanent access to these groups, even when not logged in on the console\&.
.RE
.PP
\fBDEFAULT_HOME\fR (boolean)
.RS 4
Indicate if login is allowed if we can\'t cd to the home directory\&. Default in no\&.
.sp
If set to
\fIyes\fR, the user will login in the root (/) directory if it is not possible to cd to her home directory\&.
.RE
.PP
\fBENV_HZ\fR (string)
.RS 4
If set, it will be used to define the HZ environment variable when a user login\&. The value must be preceded by
\fIHZ=\fR\&. A common value on Linux is
\fIHZ=100\fR\&.
.RE
.PP
\fBENV_PATH\fR (string)
.RS 4
If set, it will be used to define the PATH environment variable when a regular user login\&. The value can be preceded by
\fIPATH=\fR, or a colon separated list of paths (for example
\fI/bin:/usr/bin\fR)\&. The default value is
\fIPATH=/bin:/usr/bin\fR\&.
.RE
.PP
\fBENV_SUPATH\fR (string)
.RS 4
If set, it will be used to define the PATH environment variable when the superuser login\&. The value can be preceded by
\fIPATH=\fR, or a colon separated list of paths (for example
\fI/sbin:/bin:/usr/sbin:/usr/bin\fR)\&. The default value is
\fIPATH=/bin:/usr/bin\fR\&.
.RE
.PP
\fBENV_TZ\fR (string)
.RS 4
If set, it will be used to define the TZ environment variable when a user login\&. The value can be the name of a timezone preceded by
\fITZ=\fR
(for example
\fITZ=CST6CDT\fR), or the full path to the file containing the timezone specification (for example
/etc/tzname)\&.
.sp
If a full path is specified but the file does not exist or cannot be read, the default is to use
\fITZ=CST6CDT\fR\&.
.RE
.PP
\fBENVIRON_FILE\fR (string)
.RS 4
If this file exists and is readable, login environment will be read from it\&. Every line should be in the form name=value\&.
.RE
.PP
\fBERASECHAR\fR (number)
.RS 4
Terminal ERASE character (\fI010\fR
= backspace,
\fI0177\fR
= DEL)\&.
.sp
The value can be prefixed "0" for an octal value, or "0x" for an hexadecimal value\&.
.RE
.PP
\fBFAIL_DELAY\fR (number)
.RS 4
Delay in seconds before being allowed another attempt after a login failure\&.
.RE
.PP
\fBFAILLOG_ENAB\fR (boolean)
.RS 4
Enable logging and display of
/var/log/faillog
login failure info\&.
.RE
.PP
\fBFAKE_SHELL\fR (string)
.RS 4
If set,
\fBlogin\fR
will execute this shell instead of the users\' shell specified in
/etc/passwd\&.
.RE
.PP
\fBFTMP_FILE\fR (string)
.RS 4
If defined, login failures will be logged in this file in a utmp format\&.
.RE
.PP
\fBHUSHLOGIN_FILE\fR (string)
.RS 4
If defined, this file can inhibit all the usual chatter during the login sequence\&. If a full pathname is specified, then hushed mode will be enabled if the user\'s name or shell are found in the file\&. If not a full pathname, then hushed mode will be enabled if the file exists in the user\'s home directory\&.
.RE
.PP
\fBISSUE_FILE\fR (string)
.RS 4
If defined, this file will be displayed before each login prompt\&.
.RE
.PP
\fBKILLCHAR\fR (number)
.RS 4
Terminal KILL character (\fI025\fR
= CTRL/U)\&.
.sp
The value can be prefixed "0" for an octal value, or "0x" for an hexadecimal value\&.
.RE
.PP
\fBLASTLOG_ENAB\fR (boolean)
.RS 4
Enable logging and display of /var/log/lastlog login time info\&.
.RE
.PP
\fBLOGIN_RETRIES\fR (number)
.RS 4
Maximum number of login retries in case of bad password\&.
.RE
.PP
\fBLOGIN_STRING\fR (string)
.RS 4
The string used for prompting a password\&. The default is to use "Password: ", or a translation of that string\&. If you set this variable, the prompt will no be translated\&.
.sp
If the string contains
\fI%s\fR, this will be replaced by the user\'s name\&.
.RE
.PP
\fBLOGIN_TIMEOUT\fR (number)
.RS 4
Max time in seconds for login\&.
.RE
.PP
\fBLOG_OK_LOGINS\fR (boolean)
.RS 4
Enable logging of successful logins\&.
.RE
.PP
\fBLOG_UNKFAIL_ENAB\fR (boolean)
.RS 4
Enable display of unknown usernames when login failures are recorded\&.
.sp
Note: logging unknown usernames may be a security issue if an user enter her password instead of her login name\&.
.RE
.PP
\fBMAIL_CHECK_ENAB\fR (boolean)
.RS 4
Enable checking and display of mailbox status upon login\&.
.sp
You should disable it if the shell startup files already check for mail ("mailx \-e" or equivalent)\&.
.RE
.PP
\fBMAIL_DIR\fR (string)
.RS 4
The mail spool directory\&. This is needed to manipulate the mailbox when its corresponding user account is modified or deleted\&. If not specified, a compile\-time default is used\&.
.RE
.PP
\fBMAIL_FILE\fR (string)
.RS 4
Defines the location of the users mail spool files relatively to their home directory\&.
.RE
.PP
The
\fBMAIL_DIR\fR
and
\fBMAIL_FILE\fR
variables are used by
\fBuseradd\fR,
\fBusermod\fR, and
\fBuserdel\fR
to create, move, or delete the user\'s mail spool\&.
.PP
If
\fBMAIL_CHECK_ENAB\fR
is set to
\fIyes\fR, they are also used to define the
\fBMAIL\fR
environment variable\&.
.PP
\fBMOTD_FILE\fR (string)
.RS 4
If defined, ":" delimited list of "message of the day" files to be displayed upon login\&.
.RE
.PP
\fBNOLOGINS_FILE\fR (string)
.RS 4
If defined, name of file whose presence will inhibit non\-root logins\&. The contents of this file should be a message indicating why logins are inhibited\&.
.RE
.PP
\fBPORTTIME_CHECKS_ENAB\fR (boolean)
.RS 4
Enable checking of time restrictions specified in /etc/porttime\&.
.RE
.PP
\fBQUOTAS_ENAB\fR (boolean)
.RS 4
Enable setting of ulimit, umask, and niceness from passwd gecos field\&.
.RE
.PP
\fBTTYGROUP\fR (string), \fBTTYPERM\fR (string)
.RS 4
The terminal permissions: the login tty will be owned by the
\fBTTYGROUP\fR
group, and the permissions will be set to
\fBTTYPERM\fR\&.
.sp
By default, the ownership of the terminal is set to the user\'s primary group and the permissions are set to
\fI0600\fR\&.
.sp

\fBTTYGROUP\fR
can be either the name of a group or a numeric group identifier\&.
.sp
If you have a
\fBwrite\fR
program which is "setgid" to a special group which owns the terminals, define TTYGROUP to the group number and TTYPERM to 0620\&. Otherwise leave TTYGROUP commented out and assign TTYPERM to either 622 or 600\&.
.RE
.PP
\fBTTYTYPE_FILE\fR (string)
.RS 4
If defined, file which maps tty line to TERM environment parameter\&. Each line of the file is in a format something like "vt100 tty01"\&.
.RE
.PP
\fBULIMIT\fR (number)
.RS 4
Default
\fBulimit\fR
value\&.
.RE
.PP
\fBUMASK\fR (number)
.RS 4
The file mode creation mask is initialized to this value\&. If not specified, the mask will be initialized to 022\&.
.sp

\fBuseradd\fR
and
\fBnewusers\fR
use this mask to set the mode of the home directory they create
.sp
It is also used by
\fBlogin\fR
to define users\' initial umask\&. Note that this mask can be overriden by the user\'s GECOS line (if
\fBQUOTAS_ENAB\fR
is set) or by the specification of a limit with the
\fIK\fR
identifier in
\fBlimits\fR(5)\&.
.RE
.PP
\fBUSERGROUPS_ENAB\fR (boolean)
.RS 4
Enable setting of the umask group bits to be the same as owner bits (examples: 022 \-> 002, 077 \-> 007) for non\-root users, if the uid is the same as gid, and username is the same as the primary group name\&.
.sp
If set to
\fIyes\fR,
\fBuserdel\fR
will remove the user\'s group if it contains no more members, and
\fBuseradd\fR
will create by default a group with the name of the user\&.
.RE
.SH "FILES"
.PP
/var/run/utmp
.RS 4
List of current login sessions\&.
.RE
.PP
/var/log/wtmp
.RS 4
List of previous login sessions\&.
.RE
.PP
/etc/passwd
.RS 4
User account information\&.
.RE
.PP
/etc/shadow
.RS 4
Secure user account information\&.
.RE
.PP
/etc/motd
.RS 4
System message of the day file\&.
.RE
.PP
/etc/nologin
.RS 4
Prevent non\-root users from logging in\&.
.RE
.PP
/etc/ttytype
.RS 4
List of terminal types\&.
.RE
.PP
$HOME/\&.hushlogin
.RS 4
Suppress printing of system messages\&.
.RE
.SH "SEE ALSO"
.PP

\fBmail\fR(1),
\fBpasswd\fR(1),
\fBsh\fR(1),
\fBsu\fR(1),
\fBlogin.defs\fR(5),
\fBnologin\fR(5),
\fBpasswd\fR(5),
\fBsecuretty\fR(5),
\fBgetty\fR(8)\&.


src/tinylogin-0.80/login.c
/* vi: set sw=4 ts=4: */
#include "tinylogin.h"

#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <termios.h>
#include <unistd.h>
#include <utmp.h>
#include <sys/resource.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <ctype.h>
#include <time.h>
//Richard Chang 20070827 add for onlineuser
#define STATISTICS      1
#ifdef STATISTICS
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <userManager.h>
#include <debug.h>
#include <qLogEngine.h>
#endif /* STATISTICS */
#include <Util.h>
#ifdef QNAPNAS
#include <user_grp.h>
#endif

// login defines
#define LOGIN_PROMPT "\n%s login: "


/* From env.c */
extern char **newenvp;
extern size_t newenvc;


/* Stuff global to this file */
struct utmp utent;
struct passwd pwent;

static int su_mode = 0;

static void check_nologin();

#if defined TLG_FEATURE_SECURETTY
static int check_tty(const char *tty);
#else
/* #define check_tty(foo) 0 */
#define check_tty(foo) 1
#endif

static int is_my_tty();
static void login_prompt();
static void motd();
static int pw_auth();
static int set_uid_gid();

static void alarm_handler()
{
	error_msg("\nLogin timed out after %d seconds.\n", TIMEOUT);
	exit(EXIT_SUCCESS);
}

//Richard Chang 20071105 add
#define TELNET_STATE "/root/.telnet_state"
#define TELNET_USER_NAME "User Name"
#define TELNET_USER_IP "IP"
#define TELNET_USER_STATE "STATE"

int write_telnet_ip(int pid, char *user_IP)
{
	char telnet_pid[16] = "0";
	FILE		*fp = NULL;
	struct stat 	fconf;
	int		fstat = 0;
	
	fstat=stat(TELNET_STATE, &fconf);
	if (fstat==-1)
	{
		fp=fopen(TELNET_STATE, "w+");
		fclose(fp);
	}
	sprintf(telnet_pid, "%d", pid);
	WritePrivateProfileString(telnet_pid, TELNET_USER_IP, user_IP, TELNET_STATE); 
	return 0;
}

int get_telnet_ip(int pid, char *user_IP)
{
	char telnet_pid[16] = "0";
	char buf[32] = "";
	sprintf(telnet_pid, "%d", pid);
	GetPrivateProfileString(telnet_pid, TELNET_USER_IP, "0.0.0.0", buf, sizeof(buf), TELNET_STATE);
	strcpy(user_IP, buf);
	return 0;
}

int write_telnet_state(int pid, char *user_name,  char *user_state)
{
	char telnet_pid[16] = "0";
	FILE		*fp = NULL;
	struct stat 	fconf;
	int		fstat = 0;
	
	fstat=stat(TELNET_STATE, &fconf);
	if (fstat==-1)
	{
		fp=fopen(TELNET_STATE, "w+");
		fclose(fp);
	}
	sprintf(telnet_pid, "%d", pid);
	WritePrivateProfileString(telnet_pid, TELNET_USER_NAME, user_name, TELNET_STATE);
	WritePrivateProfileString(telnet_pid, TELNET_USER_STATE, user_state, TELNET_STATE);
	return 0;
}

int get_telnet__state(int pid, char *user_name,  char *user_state)
{
	char telnet_pid[16] = "0";
	char buf[128] = "";
	
	sprintf(telnet_pid, "%d", pid);
	GetPrivateProfileString(telnet_pid, TELNET_USER_NAME, "---", buf, sizeof(buf), TELNET_STATE);
	strcpy(user_name, buf);
	GetPrivateProfileString(telnet_pid, TELNET_USER_STATE, "---", buf, sizeof(buf), TELNET_STATE);
	strcpy(user_state, buf);
	return 0;
}
//end
//Richard Chang 20070827 add for onlineuser
#ifdef STATISTICS
int add_telnet_user(char *username, char *user_IP)
{
	USER_INFO user;
	user.pid=getpid();
	get_telnet_ip(user.pid, user_IP);
	strcpy(user.username, username);
	strcpy(user.ip, user_IP);
	strcpy(user.sharename, "---");
	user.app_id=shm_get_app_id("sh");
	user.schedule=SCHEDULE_PID;
	shm_add_user(&user);
	return user.pid;
}
#endif
//Richard Chang 20070906 add log
static void add_loginsucc_log(char* user_name, char *user_IP, int pid)
{
	write_telnet_state(pid, user_name,  "success");
	SendConnToLogEngine(EVENT_TYPE_INFO, user_name, user_IP, 
			"---", CONN_SERV_TELNET, CONN_ACTION_LOGIN_SUCC,  "---");
}

static void add_loginfail_log(char* user_name, char *user_IP, int pid)
{
	USER_INFO user;
	user.pid=getpid();
	pid = user.pid;
	get_telnet_ip(pid, user_IP);
	write_telnet_state(pid, user_name,  "fail");
	SendConnToLogEngine(EVENT_TYPE_WARN, user_name, user_IP, 
		"---", CONN_SERV_TELNET, CONN_ACTION_LOGIN_FAIL,  "---");
}
//end

extern int login_main(int argc, char **argv)
{
	char name[32];
	char tty[BUFSIZ];
	char full_tty[200];
	char fromhost[512];
	char *host = "";
	char *cp = NULL;
	char *tmp;
	int amroot;
	int flag;
	int fflg = 0, hflg = 0, pflg = 0;
	int failed;
	int count=0;
	struct passwd *pwd;
	time_t start, now;
	int pid = 0;
	char userIP[32] = "---";
	char user_name[32];
	FILE *readfp =NULL;
#ifdef TLG_FEATURE_SHADOWPASSWDS
	struct spwd *spwd = NULL;
#endif							/* TLG_FEATURE_SHADOWPASSWDS */
	char **envp = __environ;

	initenv();
	name[0] = '\0';
	amroot = (getuid() == 0);
	signal(SIGALRM, alarm_handler);
	while ((flag = getopt(argc, argv, "f:h:d:p")) != EOF) {
		switch (flag) {
		case 'p':
			pflg++;
			break;
		case 'f':
			/*
			 * username must be a seperate token
			 * (-f root, *NOT* -froot). --marekm
			 */
			if (optarg != argv[optind - 1]) {
				usage(login_usage);
			}
			if (!amroot) {		/* Auth bypass only if real UID is zero */
				error_msg_and_die("login: -f permission denied\n");
			}
			fflg++;
			STRFCPY(name, optarg);
			break;
		case 'h':
			hflg++;
			host = optarg;
			break;
		default:
			usage(login_usage);
		}
	}
	if (!isatty(0) || !isatty(1) || !isatty(2)) {
		exit(EXIT_FAILURE);		/* Must be a terminal */
	}

	/* XXX su_mode */
	if (!su_mode) {
		checkutmp(!amroot);
	}

	tmp = ttyname(0);
	if (tmp == NULL)
		STRFCPY(tty, "UNKNOWN");
	else {
		if (!strncmp(tmp, "/dev/", 5)) {
			STRFCPY(tty, tmp + 5);
		} else
			STRFCPY(tty, "UNKNOWN");
	}
	if (amroot) {
		bzero(utent.ut_host, sizeof utent.ut_host);
	}
	if (hflg) {
		strncpy(utent.ut_host, host, sizeof(utent.ut_host));
		cp = host;
	}
	openlog("login", LOG_PID | LOG_CONS | LOG_NOWAIT, LOG_AUTH);
	if (pflg) {
		while (*envp) {
			addenv(*envp++, NULL);
		}
	}
	if (!pflg && (tmp = getenv("TERM"))) {
		addenv("TERM", tmp);
	}
	if (optind < argc) {
		STRFCPY(name, argv[optind]);
		optind++;
	}
	if (optind < argc) {		// Set command line variables
		set_env(argc - optind, &argv[optind]);
	}
	if (cp != NULL) {
		snprintf(fromhost, sizeof(fromhost), " on `%.100s' from `%.200s'",
				 tty, cp);
	} else {
		snprintf(fromhost, sizeof(fromhost), " on `%.100s'", tty);
	}
	if (TIMEOUT > 0) {
		alarm(TIMEOUT);
	}
	__environ = newenvp;

	while (count<3) {
		failed = 0;
		if (!name[0]) {
			login_prompt(LOGIN_PROMPT, name, sizeof name);
		}
		if (!(pwd = tlg_getpwnam(name))) {
			pwent.pw_name = name;
			pwent.pw_passwd = "!";
			pwent.pw_shell = "/bin/sh";
			fflg = 0;
			failed = 1;
		} else {
			pwent = *pwd;
		}
// add by Kent 2003/01/30
#ifdef QNAPNAS
	if (strcasecmp(name, "admin")){
		//Richard Chang 20071105 add log
		add_loginfail_log(name, userIP, pid);
		error_msg_and_die("STOP !!!!!\n");
		failed = 1;
	}
#endif
// end
#ifdef TLG_FEATURE_SHADOWPASSWDS
		spwd = NULL;
		if (pwd && ((strcmp(pwd->pw_passwd, "x") == 0)
					|| (strcmp(pwd->pw_passwd, "*") == 0))) {
			spwd = getspwnam(name);
			if (spwd) {
				pwent.pw_passwd = spwd->sp_pwdp;
			} else {
				error_msg_and_die("no shadow password for `%s'%s\n", name,
								  fromhost);
			}
		}
#endif							/* TLG_FEATURE_SHADOWPASSWDS */
		if (pwent.pw_passwd[0] == '!' || pwent.pw_passwd[0] == '*') {
			failed = 1;
		}
		if (fflg) {
			fflg--;
			goto auth_ok;
		}

		/* If already root and su'ing don't ask for a password */
		if (amroot && su_mode)
			goto auth_ok;

		/* Don't check /etc/securetty if su'ing. */
		if (!su_mode && (pwent.pw_uid == 0) && (!check_tty(tty))) {
			failed = 1;
		}
		if (pwent.pw_passwd[0] == '\0') {
			goto auth_ok;
		}
		if (pw_auth(pwent.pw_passwd, name) == 0) {
#ifdef QNAPNAS
			if(Is_User_Password_Expiry(name)){
				add_loginfail_log(name, userIP, pid);
				error_msg_and_die("Password Expired !\n");
				failed = 1;
			}else if(Is_User_Account_Expiry(name)){
				add_loginfail_log(name, userIP, pid);
				error_msg_and_die("Account Expired !\n");
				failed = 1;
			}else
#endif
			goto auth_ok;
		}
		syslog(LOG_WARNING, "invalid password for `%s'%s\n",
			   pwd ? name : "UNKNOWN", fromhost);
		failed = 1;
	  auth_ok:
		if (!failed) {
			//Richard Chang 20070827 add for onlineuser
		#ifdef STATISTICS
			pid = add_telnet_user(name, userIP);
		#endif /* STATISTICS */
			//Richard Chang 20070906 add log
			add_loginsucc_log(name, userIP, pid);
			
			readfp = popen(tty,"r");
			if(readfp != NULL){
				fgets(full_tty, 10 ,readfp);
				pclose(readfp);
			}
			
			break;
		}
		if (pwent.pw_passwd[0] == '\0') {
			pw_auth("!", name);
		}
		strcpy(user_name, name);
		bzero(name, sizeof name);
		time(&start);
		now = start;
		while (difftime(now, start) < FAIL_DELAY) {
			sleep(FAIL_DELAY);
			time(&now);
		}

		/* XXX su_mode */
		if (su_mode) {
			error_msg_and_die("su: incorrect password\n");
		} else {
			//Richard Chang 20070906 add log
			add_loginfail_log(user_name, userIP, pid);
			puts("Login incorrect");
		}
		count++;
	}
	if (count>=3)
		exit (EXIT_FAILURE);
	(void) alarm(0);
	check_nologin();
	if (getenv("IFS")) {
		addenv("IFS= \t\n", NULL);
	}
	setutmp(name, tty);
	if (*tty != '/') {
		snprintf(full_tty, sizeof full_tty, "/dev/%s", tty);
	} else {
		strncpy(full_tty, tty, sizeof full_tty);
	}
	if (!is_my_tty(full_tty)) {
		syslog(LOG_ERR, "unable to determine TTY name, got %s\n",
			   full_tty);
	}
	/* Try these, but don't complain if they fail 
	 * (for example when the root fs is read only) */
	chown(full_tty, pwent.pw_uid, pwent.pw_gid);
	chmod(full_tty, 0600);

	if (set_uid_gid() != 0) {
		exit(EXIT_FAILURE);
	}
	setup_env(&pwent);
	motd();
	signal(SIGINT, SIG_DFL);	/* default interrupt signal */
	signal(SIGQUIT, SIG_DFL);	/* default quit signal */
	signal(SIGTERM, SIG_DFL);	/* default terminate signal */
	signal(SIGALRM, SIG_DFL);	/* default alarm signal */
	signal(SIGHUP, SIG_DFL);	/* added this.  --marekm */
	tlg_endpwent();					/* stop access to password file */
	tlg_endgrent();					/* stop access to group file */
#ifdef TLG_FEATURE_SHADOWPASSWDS
	endspent();					/* stop access to shadow passwd file */
//  endsgent();                    /* stop access to shadow group file */
#endif							/* TLG_FEATURE_SHADOWPASSWDS */
	if (pwent.pw_uid == 0) {
		syslog(LOG_INFO, "root login %s\n", fromhost);
	}
	closelog();
	shell(pwent.pw_shell, (char *) 0);	/* exec the shell finally. */
	 /*NOTREACHED*/ return (0);
}





static void check_nologin()
{
	if (access(NOLOGIN_FILE, F_OK) == 0) {
		FILE *nlfp;
		int c;

		if ((nlfp = fopen(NOLOGIN_FILE, "r"))) {
			while ((c = getc(nlfp)) != EOF) {
				if (c == '\n')
					putchar('\r');
				putchar(c);
			}
			fflush(stdout);
			fclose(nlfp);
		} else {
			printf("\r\nSystem closed for routine maintenance.\r\n");
		}
		if (pwent.pw_uid != 0) {
			closelog();
			exit(EXIT_SUCCESS);
		}
		printf("\r\n[Disconnect bypassed -- root login allowed.]\r\n");
	}
}

#ifdef TLG_FEATURE_SECURETTY
static int check_tty(const char *tty)
{
	FILE *fp;
	int i;
	char buf[BUFSIZ];

	if ((fp = fopen("/etc/securetty", "r")) == NULL) {
		syslog(LOG_WARNING, "cannot open securetty file.\n");
		/* return 0; */
		return 1;
	}
	while (fgets(buf, sizeof(buf), fp) != NULL) {
		for (i = strlen(buf) - 1; i >= 0; --i) {
			if (!isspace(buf[i])) {
				break;
			}
		}
		buf[++i] = '\0';
		if (buf[0] == '\0' || buf[0] == '#') {
			continue;
		}
		if (strcmp(buf, tty) == 0) {
			fclose(fp);
			/* return 0; */
			return 1;
		}
	}
	fclose(fp);
	/* return 1; */
	return 0;
}
#endif							/* TLG_FEATURE_SECURETTY */

/* returns 1 if true */
static int is_my_tty(const char *tty)
{
	struct stat by_name, by_fd;

	if (stat(tty, &by_name) || fstat(0, &by_fd)) {
		return 0;
	}
	if (by_name.st_rdev != by_fd.st_rdev) {
		return 0;
	} else {
		return 1;
	}
}

static void login_prompt(const char *prompt, char *name, int namesize)
{
	char buf[1024];
	char *cp;
	int i;
	void (*sigquit) ();

	sigquit = signal(SIGQUIT, _exit);
	if (prompt) {
		gethostname(buf, sizeof buf);
		printf(prompt, buf);
		fflush(stdout);
	}
	bzero(buf, sizeof buf);
	if (fgets(buf, sizeof buf, stdin) != buf) {
		exit(EXIT_FAILURE);
	}
	cp = strchr(buf, '\n');
	if (!cp) {
		exit(EXIT_FAILURE);
	}
	*cp = '\0';

	for (cp = buf; *cp == ' ' || *cp == '\t'; cp++);
	for (i = 0; i < namesize - 1 && isgraph(*cp); name[i++] = *cp++);
	while (isgraph(*cp)) {
		cp++;
	}
	if (*cp) {
		cp++;
	}
	name[i] = '\0';
	signal(SIGQUIT, sigquit);
}

static void motd()
{
	FILE *fp;
	register int c;

	if ((fp = fopen("/etc/motd", "r")) != NULL) {
		while ((c = getc(fp)) != EOF) {
			putchar(c);
		}
		fclose(fp);
	}
}

static int pw_auth(const char *cipher, const char *user)
{
	char *clear = NULL;
	int retval;

	if (cipher == (char *) 0 || *cipher == '\0') {
		return 0;
	}
	clear = getpass("Password: ");
	if (!clear) {
		static char c[1];

		c[0] = '\0';
		clear = c;
	}
	retval = strcmp(pw_encrypt(clear, cipher), cipher);
	bzero(clear, strlen(clear));
	return retval;
}

static int set_uid_gid()
{
	if (tlg_initgroups(pwent.pw_name, pwent.pw_gid) == -1) {
		perror("initgroups");
		syslog(LOG_ERR, "initgroups failed for user `%s': %m\n",
			   pwent.pw_name);
		closelog();
		/* return -1; */
		return 1;
	}
	if (setgid(pwent.pw_gid) == -1) {
		perror("setgid");
		syslog(LOG_ERR, "bad group ID `%d' for user `%s': %m\n",
			   pwent.pw_gid, pwent.pw_name);
		closelog();
		/* return -1; */
		return 1;
	}
	if (setuid(pwent.pw_uid)) {
		perror("setuid");
		syslog(LOG_ERR, "bad user ID `%d' for user `%s': %m\n",
			   pwent.pw_uid, pwent.pw_name);
		closelog();
		/* return -1; */
		return 1;
	}
	return 0;
}


/* + Construct an argv that login_main() can parse.
 * + For options that can't be specified on argv,
 *   modify global variables.  (ewwww)
 * + return argc
 */
static int
construct_argv(
			   char **argv,
			   char *username, int preserve, char *shell, char *command)
{
	int argc = 0;

	argv[argc++] = "su";
	if (preserve) {
		argv[argc++] = "-p";
	}
	argv[argc++] = username;

	argv[argc] = NULL;
	return argc;
}


/* 
 * TODO : I need to see if I can support the lone dash option.
 *      : I need to try to support the other options
 *      : -* poor schizophrenic login_main() *-
 `.________________________________________________________________________ */
int su_main(int argc, char **argv)
{
	int flag;
	int opt_preserve = 0;
	int opt_loginshell = 0;
	char *opt_shell = NULL;
	char *opt_command = NULL;
	char *username = "root";

	char *custom_argv[16];
	int custom_argc;

	su_mode = 1;

	/* su called w/ no args */
	if (argc == 1) {
		custom_argc = construct_argv(custom_argv,
									 username,
									 opt_preserve, opt_shell, opt_command);
		return login_main(custom_argc, custom_argv);
	}

	/* getopt */
	while ((flag = getopt(argc, argv, "c:hmps:")) != EOF) {
		switch (flag) {

		case 'c':
			opt_command = optarg;
			break;

		case 'h':
			usage(su_usage);
			break;

		case 'm':
		case 'p':
			opt_preserve = 1;
			break;

		case 's':
			opt_shell = optarg;
			break;

		default:
			usage(su_usage);
			break;
		}
	}

	/* get user if specified */
	if (optind < argc) {
		if (strcmp(argv[optind], "-") == 0) {
			opt_loginshell = 1;
			if ((optind + 1) < argc) {
				username = argv[++optind];
			}
		} else {
			username = argv[optind];
		}
		optind++;
	}

	/* construct custom_argv */
	custom_argc = construct_argv(custom_argv,
								 username,
								 opt_preserve, opt_shell, opt_command);

	/* reset getopt | how close to kosher is this? | damn globals */
	optind = 0;

	return login_main(custom_argc, custom_argv);
}
